File: hackerrank/matrix/diagdiff1.rb
#!/bin/ruby

## Solution to https://www.hackerrank.com/challenges/diagonal-difference/problem

require 'json'
require 'stringio'

# Complete the diagonalDifference function below.
def diagonalDifference(arr)
    # Debug 
    #p "Arr.length()=#{arr.length}" 
    d1 = 0 
    d2 = 0 
    for i in 0 .. arr.length-1
        d1 += arr[i][i]
        d2 += arr[i][arr.length-1-i]
    end
    return (d1-d2).abs
    
end

fptr = File.open(ENV['OUTPUT_PATH'], 'w')

n = gets.to_i

arr = Array.new(n)

n.times do |i|
    arr[i] = gets.rstrip.split(' ').map(&:to_i)
end

result = diagonalDifference arr

fptr.write result
fptr.write "\n"

fptr.close()

File: hackerrank/matrix/diagdiff1.py
#!/bin/python

## Solution to https://www.hackerrank.com/challenges/diagonal-difference/problem

import math
import os
import random
import re
import sys

# The O(N) solution 
# Beside the obvious syntactic differences the main interesting one is the array length is not represented as a container property but it is returned by the keyword len() which represents the general way to get the monodimensisonal container number of element 
def diagonalDifference(arr):
    d1=0
    d2=0
    
    for i in range(len(arr)):
        d1 += arr[i][i]
        d2 += arr[i][len(arr)-1-i]
    return abs(d1-d2)
    
if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(raw_input())

    arr = []

    for _ in xrange(n):
        arr.append(map(int, raw_input().rstrip().split()))

    result = diagonalDifference(arr)

    fptr.write(str(result) + '\n')

    fptr.close()


